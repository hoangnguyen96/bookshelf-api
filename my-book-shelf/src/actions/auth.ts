"use server";

import { AuthError } from "next-auth";
import { signIn, signOut } from "@app/auth";
import { ERROR_MESSAGES } from "@app/constants";
import { User } from "@app/models";

export const authenticate = async (formData: Partial<User>) => {
  console.log("-form:", formData);

  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      switch (error.type) {
        case "CredentialsSignin":
          return ERROR_MESSAGES.EMAIL_PASSWORD_INVALID;
        default:
          return "Something went wrong.";
      }
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
