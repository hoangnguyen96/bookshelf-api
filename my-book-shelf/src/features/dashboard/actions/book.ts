import { BookType } from "@app/models";
import { API_ROUTES, MESSAGES } from "@app/constants";
import { api } from "@app/services";
import { dividePaginationBooks, getTopBook } from "@app/utils";

export const getAllBook = async () => {
  try {
    const data = await api.get<BookType[]>(API_ROUTES.BOOKS);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getPaginatedBook = async (params?: string) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}`,
      {
        cache: "no-store",
      }
    );
    const result = dividePaginationBooks(data);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBooksByLimit = async (params?: string, limit?: number) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}page=1`,
      { next: { revalidate: 60 } }
    );

    const result = getTopBook(data, limit);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookByParams = async (params: string) => {
  try {
    const data = await api.get<BookType[]>(`${API_ROUTES.BOOKS}?${params}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookById = async (id: string) => {
  try {
    const data = await api.get<BookType>(`${API_ROUTES.BOOKS}/${parseInt(id)}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const addBook = async (payload: Partial<BookType>) => {
  return await api.post(API_ROUTES.BOOKS, payload);
};

export const updateBookById = async (
  id: string,
  payload: Partial<BookType>
) => {
  return await api.put(`${API_ROUTES.BOOKS}/${id}`, payload);
};

export const deleteBook = async (id: string) => {
  return await api.delete(`${API_ROUTES.BOOKS}/${id}`);
};
