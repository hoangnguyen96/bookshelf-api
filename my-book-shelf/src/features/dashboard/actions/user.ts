import { API_ROUTES } from "@app/constants";
import { User } from "@app/models";
import { api } from "@app/services";

export const getAllUser = async () => {
  const data = await api.get<User[]>(API_ROUTES.USER);

  return data || [];
};

export const getUserById = async (id: string) => {
  try {
    const data = await api.get<User[]>(`${API_ROUTES.USER}?userId=${id}`);

    return data[0];
  } catch (error) {
    return [];
  }
};

export const getUserByEmail = async (email: string) => {
  try {
    const user = (await api.get(`${API_ROUTES.USER}?email=${email}`)) as User[];

    return user;
  } catch (error) {
    return [];
  }
};

export const updateUserById = async (id: string, payload: Partial<User>) => {
  return await api.put(`${API_ROUTES.USER}/${id}`, payload);
};

export const addUser = async (payload: Partial<User>) => {
  return await api.post(API_ROUTES.USER, payload);
};
