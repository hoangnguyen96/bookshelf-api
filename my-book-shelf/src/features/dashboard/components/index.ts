export * from "./home-list";
export * from "./home-list-by-params";
export * from "./search-list";
export * from "./search-list-by-params";
export * from "./my-book-shelf";
export * from "./my-book-shelf-by-params";
export * from "./my-book-shelf-favorites";
export * from "./my-book-shelf-favorites-by-params";
export * from "./preview-book-details";
export * from "./profile-update";
export * from "./contribute";
export * from "./contribute-update";
export * from "./contribute-complete";
export * from "./contribute-list";
export * from "./list-top-contribute";
