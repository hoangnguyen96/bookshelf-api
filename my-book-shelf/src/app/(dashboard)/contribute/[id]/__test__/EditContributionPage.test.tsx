import { render, waitFor } from "@testing-library/react";
import { DATA_BOOKS } from "@app/mocks/data";
import { getBookById } from "@app/features/dashboard/actions";
import EditContributionPage from "../page";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  notFound: jest.fn(),
}));

jest.mock("@app/features/dashboard/actions", () => ({
  getBookById: jest.fn(),
}));

describe("Contribute Update", () => {
  const mockBook = DATA_BOOKS[0];

  beforeEach(() => {
    jest.clearAllMocks();
    (getBookById as jest.Mock).mockResolvedValue(mockBook);
  });

  it("Should render correctly snapshot", async () => {
    const { container } = render(
      await EditContributionPage({ params: { id: "1" } })
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
