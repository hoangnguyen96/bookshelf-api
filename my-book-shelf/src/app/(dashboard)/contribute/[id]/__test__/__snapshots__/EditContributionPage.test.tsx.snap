// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contribute Update Should render correctly snapshot 1`] = `<div />`;
